contract/src/internals/helpers.rs

```diff
@@ -50,6 +88,114 @@ impl Keypom {
         );
     }
 
+    /// Internal function to assert that the predecessor is the contract owner
+    pub(crate) fn assert_claim_timestamps(&mut self, drop_id: DropId, drop: &mut Drop, key_info: &mut KeyInfo, signer_pk: &PublicKey) -> bool {
+        // Ensure enough time has passed if a start timestamp was specified in the config.
+        let current_timestamp = env::block_timestamp();
+        let desired_start_timestamp = drop
+            .config
+            .clone()
+            .and_then(|c| c.start_timestamp)
+            .unwrap_or(0);
+
+        if current_timestamp < desired_start_timestamp {
+            let used_gas = env::used_gas();
+
+            let amount_to_decrement =
+                (used_gas.0 + GAS_FOR_PANIC_OFFSET.0) as u128 * self.yocto_per_gas;
+            near_sdk::log!("Drop isn't claimable until {}. Current timestamp is {}. Decrementing allowance by {}. Used GAS: {}", desired_start_timestamp, current_timestamp, amount_to_decrement, used_gas.0);
+
+            key_info.allowance -= amount_to_decrement;
+            near_sdk::log!("Allowance is now {}", key_info.allowance);
+            drop.pks.insert(&signer_pk, &key_info);
+            self.drop_for_id.insert(&drop_id, &drop);
+            return false;
+        }
+
+        // Ensure the end timestamp hasn't passed and the key is still usable
+        let desired_end_timestamp = drop
+            .config
+            .clone()
+            .and_then(|c| c.end_timestamp)
+            .unwrap_or(u64::MAX);
+        if current_timestamp > desired_end_timestamp {
+            let used_gas = env::used_gas();
+
+            let amount_to_decrement =
+                (used_gas.0 + GAS_FOR_PANIC_OFFSET.0) as u128 * self.yocto_per_gas;
+            near_sdk::log!("Drop claim period passed. Claimable up until {}. Current timestamp is {}. Decrementing allowance by {}. Used GAS: {}", desired_end_timestamp, current_timestamp, amount_to_decrement, used_gas.0);
+
+            key_info.allowance -= amount_to_decrement;
+            near_sdk::log!("Allowance is now {}", key_info.allowance);
+            drop.pks.insert(&signer_pk, &key_info);
+            self.drop_for_id.insert(&drop_id, &drop);
```

Comment:
This could be a helper. Similar pattern shows up next two blocks. Seems like it could be parameterized.